index.js

```diff
@@ -0,0 +1,111 @@
+'use strict'
+
+const {
+  extractSocketDetails,
+  protocolDecoder
+} = require('aedes-protocol-decoder')
+const http = require('http')
+const https = require('https')
+const http2 = require('http2')
+const net = require('net')
+const tls = require('tls')
+const WebSocket = require('ws')
+
+const defaultOptions = {
+  ws: null,
+  http: null,
+  https: null,
+  http2: null,
+  tls: null,
+  trustProxy: false,
+  serverFactory: null,
+  protocolDecoder: protocolDecoder,
+  extractSocketDetails: extractSocketDetails
+}
+
+const createServer = (aedes, options) => {
+  if (!aedes || !aedes.handle) {
+    throw new Error('Missing aedes handler')
+  }
+
+  options = Object.assign({}, defaultOptions, options)
+
+  let server = null
+  if (options.serverFactory) {
+    server = options.serverFactory(aedes, options)
+  } else if (options.tls) {
+    server = tls.createServer(options.tls, (conn) => {
+      bindConnection(aedes, options, conn)
+    })
+  } else if (options.ws) {
+    if (options.https) {
+      if (options.http2) {
+        server = http2.createSecureServer({
+          ...options.http2,
+          ...options.https
+        })
+      } else {
+        server = https.createServer({
+          ...(options.http || {}),
+          ...options.https
+        })
+      }
+    } else {
+      if (options.http2) {
+        server = http2.createServer(options.http2)
+      } else {
+        server = http.createServer(options.http || {})
+      }
+    }
+    const ws = new WebSocket.Server({ server })
+    ws.on('connection', (conn, req) => {
+      const stream = WebSocket.createWebSocketStream(conn)
+      // the _socket object is needed in bindConnection to retrieve info from the stream
+      // before passing it to aedes.handle
+      stream._socket = conn._socket
+      bindConnection(aedes, options, stream, req)
+    })
+  } else {
+    server = net.createServer((conn) => {
+      bindConnection(aedes, options, conn)
+    })
+  }
+  return server
+}
+
+const bindConnection = (aedes, options, conn, req = {}) => {
+  if (options.trustProxy) {
+    extractConnectionDetails(aedes, options, conn, req)
+  } else {
+    req.connDetails = options.extractSocketDetails(conn.socket || conn)
+    aedes.handle(conn, req)
+  }
+}
+
+const extractConnectionDetails = (aedes, options, conn, req = {}) => {
+  const onReadable = () => {
+    // buffer should contain the whole proxy header if any
+    // see https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+    const buffer = conn.read(null)
+    if (buffer) {
+      const protocol = options.protocolDecoder(conn, buffer, req)
+      req.connDetails = protocol
+      conn.removeListener('readable', onReadable)
```

Comment:
this should also remove the `error` listener.

Reply:
Right, then it makes me think should we remove the `readable` listener in case of `error` ?

Reply:
Yes, exactly. It's not strictly necessary as after error the stream is useless - there are no more expected chunks.

Reply:
Thanks for the information, i will remove that listener for consistency.

Reply:
Let me know when this is ready :)